src/manager.rs

```diff
@@ -0,0 +1,532 @@
+                }
+                Value::Array(ref mut vec) => {
+                    if let Ok(x) = token.parse::<usize>() {
+                        if is_last {
+                            let v = std::mem::replace(&mut vec[x], Value::Null);
+                            if let Some(res) = (func)(v)? {
+                                vec[x] = res;
+                            } else {
+                                vec.remove(x);
+                            }
+                            return Ok(());
+                        }
+                        vec.get_mut(x)
+                    } else {
+                        None
+                    }
+                }
+                _ => None,
+            };
+
+            if let Some(t) = target_opt {
+                target = t;
+            } else {
+                break;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn do_op<F>(&mut self, paths: Vec<String>, mut op_fun: F) -> Result<(), RedisError>
+    where
+        F: FnMut(Value) -> Result<Option<Value>, Error>,
+    {
+        if paths.len() == 0 {
+            // updating the root require special treatment
+            let root = self.get_value().unwrap().unwrap();
+            let res = (op_fun)(root.take())?;
+            self.set_root(res)?;
+        } else {
+            self.update(&paths, self.get_value().unwrap().unwrap(), op_fun)?;
+        }
+
+        Ok(())
+    }
+
+    fn do_num_op<F1, F2>(
+        &mut self,
+        path: Vec<String>,
+        num: &str,
+        mut op1_fun: F1,
+        mut op2_fun: F2,
+    ) -> Result<Number, RedisError>
+    where
+        F1: FnMut(i64, i64) -> i64,
+        F2: FnMut(f64, f64) -> f64,
+    {
+        let in_value = &serde_json::from_str(num)?;
+        if let Value::Number(in_value) = in_value {
+            let mut res = None;
+            self.do_op(path, |v| {
+                let num_res = match (v.as_i64(), in_value.as_i64()) {
+                    (Some(num1), Some(num2)) => ((op1_fun)(num1, num2)).into(),
+                    _ => {
+                        let num1 = v.as_f64().unwrap();
+                        let num2 = in_value.as_f64().unwrap();
+                        Number::from_f64((op2_fun)(num1, num2)).unwrap()
+                    }
+                };
+                res = Some(Value::Number(num_res));
+                Ok(res.clone())
+            })?;
+            match res {
+                None => Err(RedisError::Str("path does not exists")),
+                Some(n) => match n {
+                    Value::Number(n) => Ok(n),
+                    _ => Err(RedisError::Str("return value is not a number")),
+                },
+            }
+        } else {
+            Err(RedisError::Str("bad input number"))
+        }
+    }
+
+    fn set_root(&mut self, v: Option<Value>) -> Result<(), RedisError> {
+        match v {
+            Some(inner) => {
+                let v = self.key.get_value::<RedisJSON>(&REDIS_JSON_TYPE)?;
```

Comment:
Calling `get_value` could be redundant here, such as when coming from `command_json_set` calling `WriteHolder::set_value` calling `set_root`, where the value was already obtained before in `command_json_set` by
`let current = redis_key.get_value()?;`